indico/modules/rb_new/client/js/common/timeline/DateNavigator.jsx

```diff
@@ -0,0 +1,257 @@
+/* This file is part of Indico.
+ * Copyright (C) 2002 - 2018 European Organization for Nuclear Research (CERN).
+ *
+ * Indico is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation; either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * Indico is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Indico; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import {Button} from 'semantic-ui-react';
+import RCCalendar from 'rc-calendar';
+import DatePicker from 'rc-calendar/lib/Picker';
+import {Translate} from 'indico/react/i18n';
+import {serializeDate, toMoment} from 'indico/utils/date';
+import {isDateWithinRange} from '../../util';
+
+/**
+ * Component that renders a 'mode selector' (day/week/month) and a date picker.
+ * This is used in timeline-style views (e.g. 'Book a Room' or 'Calendar').
+ */
+export default class DateNavigator extends React.Component {
+    static propTypes = {
+        selectedDate: PropTypes.string.isRequired,
+        mode: PropTypes.string.isRequired,
+        dateRange: PropTypes.array.isRequired,
+        onDateChange: PropTypes.func.isRequired,
+        onModeChange: PropTypes.func.isRequired,
+        isLoading: PropTypes.bool,
+        disabled: PropTypes.bool
+    };
+
+    static defaultProps = {
+        isLoading: false,
+        disabled: false
+    };
+
+    constructor(props) {
+        super(props);
+        const {mode} = this.props;
+        this.setDateWithMode(this.selectedDate, mode);
+    }
+
+    componentDidUpdate(prevProps) {
+        const {mode, selectedDate} = this.props;
+        const {mode: prevMode, selectedDate: prevSelectedDate} = prevProps;
+        if (prevMode !== mode || !prevSelectedDate === selectedDate) {
```

Comment:
Is this some kind of a shortcut that I don't know? Which does some black magic? Or is it better to use `!==`?

Reply:
Yeah, I don't know where that came from.